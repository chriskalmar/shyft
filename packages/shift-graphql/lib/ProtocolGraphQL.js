
import {
  ProtocolType,
  dataTypes,
} from 'shift-engine';

import {
  GraphQLScalarType,
} from 'graphql';

import {
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
} from 'graphql';

import {
  GraphQLBigInt,
  GraphQLJSON,
} from './dataTypes';

const {
  DataTypeID,
  DataTypeInteger,
  DataTypeBigInt,
  DataTypeFloat,
  DataTypeBoolean,
  DataTypeString,
  DataTypeJson,
} = dataTypes;



export const ProtocolGraphQL = new ProtocolType({
  name: 'ProtocolGraphQL',
  description: 'GraphQL API protocol',
  isProtocolDataType(protocolDataType) {
    return (protocolDataType instanceof GraphQLScalarType)
  }
})

export default ProtocolGraphQL


ProtocolGraphQL.addDataTypeMap(DataTypeID, GraphQLBigInt);
ProtocolGraphQL.addDataTypeMap(DataTypeInteger, GraphQLInt);
ProtocolGraphQL.addDataTypeMap(DataTypeBigInt, GraphQLBigInt);
ProtocolGraphQL.addDataTypeMap(DataTypeFloat, GraphQLFloat);
ProtocolGraphQL.addDataTypeMap(DataTypeBoolean, GraphQLBoolean);
ProtocolGraphQL.addDataTypeMap(DataTypeString, GraphQLString);
ProtocolGraphQL.addDataTypeMap(DataTypeJson, GraphQLJSON);
