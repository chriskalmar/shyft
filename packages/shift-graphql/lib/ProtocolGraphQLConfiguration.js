

import {
  ProtocolConfiguration,
} from 'shift-engine';
import {
  generateTypeName,
  generateTypeNamePlural,
  generateTypeNamePascalCase,
  generateTypeNamePluralPascalCase,
} from './util';


class ProtocolGraphQLConfiguration extends ProtocolConfiguration {

  constructor() {
    super()

    this.enableFeatures([
      'query',
      'action',
      'mutation',
      'mutationNested',
      'mutationById',
      'mutationByUniqueness',
    ], true)
  }

  generateEntityTypeName(entity) {
    return generateTypeName(entity.name)
  }

  generateEntityTypeNamePlural(entity) {
    return generateTypeNamePlural(entity.name)
  }

  generateEntityTypeNamePascalCase(entity) {
    return generateTypeNamePascalCase(entity.name)
  }

  generateEntityTypeNamePluralPascalCase(entity) {
    return generateTypeNamePluralPascalCase(entity.name)
  }

  generateReverseConnectionFieldName(sourceEntity, sourceAttributeName) {
    const typeNamePlural = this.generateEntityTypeNamePlural(sourceEntity)
    return generateTypeName(`${typeNamePlural}-by-${sourceAttributeName}`)
  }

  generateReferenceFieldName(referenceEntity, attribute) {
    const fieldName = this.generateFieldName(attribute)
    return generateTypeName(`${referenceEntity.name}-by-${fieldName}`)
  }


  generateFieldName(attribute) {
    return generateTypeName(attribute.name)
  }

  generateListQueryTypeName(entity) {
    const typeNamePlural = this.generateEntityTypeNamePlural(entity)
    return generateTypeName(`all-${typeNamePlural}`)
  }

  generateInstanceQueryTypeName(entity) {
    return this.generateEntityTypeName(entity)
  }

  generateInstanceByUniqueQueryTypeName(entity, attribute) {
    const typeName = this.generateEntityTypeName(entity)
    const fieldName = this.generateFieldName(attribute)
    return generateTypeName(`${typeName}-by-${fieldName}`)
  }


  generateMutationInstanceInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity)
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-instance-input`)
  }

  generateMutationInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity)
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-input`)
  }

  generateMutationByPrimaryAttributeInputTypeName(entity, mutation, attribute) {
    const typeName = this.generateEntityTypeName(entity)
    const fieldName = this.generateFieldName(attribute)
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-by-${fieldName}-input`)
  }


  generateUniquenessAttributesName(entity, attributes) {
    return generateTypeName(attributes.join('-and-'))
  }

  generateInstanceUniquenessInputTypeName(entity, uniquenessAttributesName) {
    const typeName = this.generateEntityTypeName(entity)
    return generateTypeNamePascalCase(`${typeName}-instance-uniqueness-on-${uniquenessAttributesName}-input`)
  }


  generateMutationInstanceNestedInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity)
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-instance-nested-input`)
  }

  generateMutationNestedInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity)
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-nested-input`)
  }


  generateMutationOutputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity)
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-output`)
  }


}


export default ProtocolGraphQLConfiguration


export const isProtocolGraphQLConfiguration = (obj) => {
  return (obj instanceof ProtocolGraphQLConfiguration)
}
