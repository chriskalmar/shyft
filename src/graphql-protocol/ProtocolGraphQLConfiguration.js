import {
  generateTypeName,
  generateTypeNamePlural,
  generateTypeNamePascalCase,
  generateTypeNamePluralPascalCase,
  generateTypeNameUpperCase,
} from './util';
import { MAX_PAGE_SIZE } from './protocolGraphqlConstants';
import { ProtocolConfiguration } from '../engine/protocol/ProtocolConfiguration';

export class ProtocolGraphQLConfiguration extends ProtocolConfiguration {
  constructor() {
    super();

    this.enableFeatures(
      [
        'query',
        'action',
        'mutation',
        'mutationNested',
        'mutationById',
        'mutationByUniqueness',
      ],
      true,
    );
  }

  getMaxPageSize() {
    return Number(process.env.MAX_PAGE_SIZE) || MAX_PAGE_SIZE;
  }

  /* name generators */

  generateEntityTypeName(entity) {
    return generateTypeName(entity.name);
  }

  generateEntityTypeNamePlural(entity) {
    return generateTypeNamePlural(entity.name);
  }

  generateEntityTypeNamePascalCase(entity) {
    return generateTypeNamePascalCase(entity.name);
  }

  generateEntityTypeNamePluralPascalCase(entity) {
    return generateTypeNamePluralPascalCase(entity.name);
  }

  generateReverseConnectionFieldName(sourceEntity, sourceAttributeName) {
    const typeNamePlural = this.generateEntityTypeNamePlural(sourceEntity);
    return generateTypeName(`${typeNamePlural}-by-${sourceAttributeName}`);
  }

  generateReferenceFieldName(referenceEntity, attribute) {
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(`${referenceEntity.name}-by-${fieldName}`);
  }

  generateFieldName(attribute) {
    return generateTypeName(attribute.name);
  }

  generateI18nJsonFieldName(attribute) {
    const fieldName = this.generateFieldName(attribute);
    return `${fieldName}_i18nJson`;
  }

  generateI18nFieldName(attribute) {
    const fieldName = this.generateFieldName(attribute);
    return `${fieldName}_i18n`;
  }

  generateI18nFieldTypeName(entity, attribute) {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeNamePascalCase(`${typeName}-${fieldName}-i18n`);
  }

  generateMutationI18nAttributeInputTypeName(entity, mutation, attribute) {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeNamePascalCase(
      `${mutation.name}-${typeName}-${fieldName}-i18n-input`,
    );
  }

  generateListQueryTypeName(entity) {
    const typeNamePlural = this.generateEntityTypeNamePlural(entity);
    return generateTypeName(`all-${typeNamePlural}`);
  }

  generateInstanceQueryTypeName(entity) {
    return this.generateEntityTypeName(entity);
  }

  generateInstanceByUniqueQueryTypeName(entity, attribute) {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(`${typeName}-by-${fieldName}`);
  }

  generateMutationInstanceInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${mutation.name}-${typeName}-instance-input`,
    );
  }

  generateMutationInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-input`);
  }

  generateMutationByPrimaryAttributeInputTypeName(entity, mutation, attribute) {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeNamePascalCase(
      `${mutation.name}-${typeName}-by-${fieldName}-input`,
    );
  }

  generateUniquenessAttributesName(entity, attributes) {
    return generateTypeName(attributes.join('-and-'));
  }

  generateUniquenessAttributesFieldName(
    entity,
    attribute,
    uniquenessAttributesName,
  ) {
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(
      `${fieldName}-by-unique-${uniquenessAttributesName}`,
    );
  }

  generateInstanceUniquenessInputTypeName(entity, uniquenessAttributesName) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${typeName}-instance-uniqueness-on-${uniquenessAttributesName}-input`,
    );
  }

  generateMutationInstanceNestedInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${mutation.name}-${typeName}-instance-nested-input`,
    );
  }

  generateMutationNestedInputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${mutation.name}-${typeName}-nested-input`,
    );
  }

  generateMutationOutputTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${mutation.name}-${typeName}-output`);
  }

  generateMutationTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeName(`${mutation.name}-${typeName}`);
  }

  generateMutationNestedTypeName(entity, mutation) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeName(`${mutation.name}-${typeName}-nested`);
  }

  generateMutationByPrimaryAttributeTypeName(entity, mutation, attribute) {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(`${mutation.name}-${typeName}-by-${fieldName}`);
  }

  generateActionTypeName(action) {
    return generateTypeName(action.name);
  }

  generateDataInputTypeName(baseName) {
    return generateTypeNamePascalCase(`${baseName}-data-input`);
  }

  generateNestedDataInputTypeName(baseName, nestedParamName, level) {
    const levelStr = level > 1 ? `L${level}` : '';

    return generateTypeNamePascalCase(
      `${baseName}-${nestedParamName}-${levelStr}-data-input`,
    );
  }

  generateInputTypeName(baseName) {
    return generateTypeNamePascalCase(`${baseName}-input`);
  }

  generateDataOutPutTypeName(baseName) {
    return generateTypeNamePascalCase(`${baseName}-data-output`);
  }

  generateNestedDataOutPutTypeName(baseName, nestedParamName, level) {
    const levelStr = level > 1 ? `L${level}` : '';

    return generateTypeNamePascalCase(
      `${baseName}-${nestedParamName}-${levelStr}-data-output`,
    );
  }

  generateOutPutTypeName(baseName) {
    return generateTypeNamePascalCase(`${baseName}-output`);
  }

  generateSortKeyName(attribute, ascending) {
    const direction = ascending ? 'ASC' : 'DESC';

    return `${generateTypeNameUpperCase(attribute.name)}_${direction}`;
  }

  generateSortInputTypeName(entity) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-order-by`);
  }

  generateFilterInputTypeName(entity) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-filter`);
  }

  generateConnectionEdgeTypeName(entity) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-edge`);
  }

  generateConnectionTypeName(entity) {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-connection`);
  }
}

export const isProtocolGraphQLConfiguration = obj => {
  return obj instanceof ProtocolGraphQLConfiguration;
};
