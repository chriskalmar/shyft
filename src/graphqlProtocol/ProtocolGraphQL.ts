import {
  GraphQLScalarType,
  GraphQLID,
  GraphQLInt,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
  GraphQLEnumType,
} from 'graphql';

import { ProtocolType } from '../engine/protocol/ProtocolType';
import {
  DataTypeID,
  DataTypeInteger,
  DataTypeBigInt,
  DataTypeFloat,
  DataTypeBoolean,
  DataTypeString,
  DataTypeJson,
  DataTypeTimestamp,
  DataTypeTimestampTz,
  DataTypeDate,
  DataTypeTime,
  DataTypeTimeTz,
  DataTypeUUID,
  DataTypeI18n,
} from '../engine/datatype/dataTypes';

import { isDataTypeState } from '../engine/datatype/DataTypeState';
import { isDataTypeEnum } from '../engine/datatype/DataTypeEnum';
import { isObjectDataType } from '../engine/datatype/ObjectDataType';
import { isListDataType } from '../engine/datatype/ListDataType';

import { graphRegistry } from './graphRegistry';

import {
  GraphQLBigInt,
  GraphQLJSON,
  GraphQLDateTime,
  GraphQLDate,
  GraphQLTime,
} from './dataTypes';

import { generateDataInput, generateDataOutput } from './io';

export const ProtocolGraphQL = new ProtocolType({
  name: 'ProtocolGraphQL',
  description: 'GraphQL API protocol',
  isProtocolDataType(protocolDataType) {
    return protocolDataType instanceof GraphQLScalarType;
  },
});

ProtocolGraphQL.addDataTypeMap(DataTypeID, GraphQLID);
ProtocolGraphQL.addDataTypeMap(DataTypeInteger, GraphQLInt);
ProtocolGraphQL.addDataTypeMap(DataTypeBigInt, GraphQLBigInt);
ProtocolGraphQL.addDataTypeMap(DataTypeFloat, GraphQLFloat);
ProtocolGraphQL.addDataTypeMap(DataTypeBoolean, GraphQLBoolean);
ProtocolGraphQL.addDataTypeMap(DataTypeString, GraphQLString);
ProtocolGraphQL.addDataTypeMap(DataTypeJson, GraphQLJSON);
ProtocolGraphQL.addDataTypeMap(DataTypeTimestamp, GraphQLDateTime);
ProtocolGraphQL.addDataTypeMap(DataTypeTimestampTz, GraphQLDateTime);
ProtocolGraphQL.addDataTypeMap(DataTypeTime, GraphQLTime);
ProtocolGraphQL.addDataTypeMap(DataTypeTimeTz, GraphQLTime);
ProtocolGraphQL.addDataTypeMap(DataTypeDate, GraphQLDate);
ProtocolGraphQL.addDataTypeMap(DataTypeUUID, GraphQLID);
ProtocolGraphQL.addDataTypeMap(DataTypeI18n, GraphQLJSON);

const dataTypesRegistry = {
  object: {},
  enum: {},
};

ProtocolGraphQL.addDynamicDataTypeMap(isDataTypeState, attributeType => {
  const name = attributeType.name;
  const values = {};

  if (dataTypesRegistry.enum[name]) {
    if (attributeType === dataTypesRegistry.enum[name].attributeType) {
      return dataTypesRegistry.enum[name].type;
    }
  }

  const states = attributeType.states;
  const stateNames = Object.keys(states);
  stateNames.map(stateName => {
    values[stateName] = {
      value: states[stateName],
    };
  });

  const type = new GraphQLEnumType({
    name,
    values,
  });

  dataTypesRegistry.enum[name] = {
    type,
    attributeType,
  };

  return type;
});

ProtocolGraphQL.addDynamicDataTypeMap(isDataTypeEnum, attributeType => {
  const name = attributeType.name;
  const values = {};

  if (dataTypesRegistry.enum[name]) {
    if (attributeType === dataTypesRegistry.enum[name].attributeType) {
      return dataTypesRegistry.enum[name].type;
    }
  }

  const enumValues = attributeType.values;
  const enumValuesNames = Object.keys(enumValues);
  enumValuesNames.map(valueName => {
    values[valueName] = {
      value: enumValues[valueName],
    };
  });

  const type = new GraphQLEnumType({
    name,
    values,
  });

  dataTypesRegistry.enum[name] = {
    type,
    attributeType,
  };

  return type;
});

ProtocolGraphQL.addDynamicDataTypeMap(
  isObjectDataType,
  (attributeType, sourceName, asInput) => {
    const name = attributeType.name;
    const uniqueName = `${name}-${sourceName}-${asInput ? 'Input' : 'Output'}`;

    if (asInput) {
      if (!dataTypesRegistry.object[uniqueName]) {
        const dataInputType = generateDataInput(
          `${sourceName}-${name}`,
          attributeType.getAttributes(),
        );
        dataTypesRegistry.object[uniqueName] = dataInputType;
      }

      return dataTypesRegistry.object[uniqueName];
    }

    if (!dataTypesRegistry.object[uniqueName]) {
      const dataOutputType = generateDataOutput(
        `${sourceName}-${name}`,
        attributeType.getAttributes(),
        graphRegistry,
      );
      dataTypesRegistry.object[uniqueName] = dataOutputType;
    }

    return dataTypesRegistry.object[uniqueName];
  },
);

ProtocolGraphQL.addDynamicDataTypeMap(
  isListDataType,
  (attributeType, sourceName, asInput) => {
    const name = attributeType.name;
    const uniqueName = `${name}-${sourceName}-${asInput ? 'Input' : 'Output'}`;

    // hack: wrap list type into an object data type and extract later that single field
    // to reuse the same input / ouput logic as with object data types
    const params = {
      wrapped: {
        type: attributeType,
      },
    };

    if (asInput) {
      if (!dataTypesRegistry.object[uniqueName]) {
        const dataInputType = generateDataInput(
          `${sourceName}-${name}`,
          params,
        ) as GraphQLInputObjectType;
        dataTypesRegistry.object[
          uniqueName
        ] = dataInputType.getFields().wrapped.type;
      }

      return dataTypesRegistry.object[uniqueName];
    }

    if (!dataTypesRegistry.object[uniqueName]) {
      const dataOutputType = generateDataOutput(
        `${sourceName}-${name}`,
        params,
        graphRegistry,
      ) as GraphQLObjectType;
      dataTypesRegistry.object[
        uniqueName
      ] = dataOutputType.getFields().wrapped.type;
    }

    return dataTypesRegistry.object[uniqueName];
  },
);
