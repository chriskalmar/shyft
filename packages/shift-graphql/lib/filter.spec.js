
import {
  splitAttributeAndFilterOperator,
  transformFilterLevel,
} from './filter'

import {
  extendModelsForGql,
} from './generator';

import {
  Entity,
  DataTypeBoolean,
  DataTypeString,
  DataTypeInteger,
} from 'shift-engine';

import ProtocolGraphQL from './ProtocolGraphQL';
import ProtocolGraphQLConfiguration from './ProtocolGraphQLConfiguration';


ProtocolGraphQL.setProtocolConfiguration(new ProtocolGraphQLConfiguration())


describe('filter', () => {

  const filteredEntity = new Entity({
    name: 'FilteredEntityName',
    description: 'Just some description',
    attributes: {
      id: {
        type: DataTypeInteger,
        description: 'ID of user',
        isPrimary: true,
      },
      firstName: {
        type: DataTypeString,
        description: 'First name'
      },
      lastName: {
        type: DataTypeString,
        description: 'Last name'
      },
      isActive: {
        type: DataTypeBoolean,
        description: 'User has been active within this month'
      },
    }
  })


  extendModelsForGql([filteredEntity])


  describe('splitAttributeAndFilterOperator', () => {


    it('should split attributes from operators', () => {

      expect(splitAttributeAndFilterOperator('test')).toEqual({ attributeName: 'test' })
      expect(splitAttributeAndFilterOperator('__test')).toEqual({ attributeName: '__test' })
      expect(splitAttributeAndFilterOperator('login__lt')).toEqual({ attributeName: 'login', operator: 'lt' })
      expect(splitAttributeAndFilterOperator('firstName__gte')).toEqual({ attributeName: 'firstName', operator: 'gte' })
      expect(splitAttributeAndFilterOperator('last_name__ne')).toEqual({ attributeName: 'last_name', operator: 'ne' })
      expect(splitAttributeAndFilterOperator('some__long_attribute__name__lte')).toEqual({ attributeName: 'some__long_attribute__name', operator: 'lte' })
      expect(splitAttributeAndFilterOperator('__some_name__ne')).toEqual({ attributeName: '__some_name', operator: 'ne' })
      expect(splitAttributeAndFilterOperator('___some_name__ne')).toEqual({ attributeName: '___some_name', operator: 'ne' })
      expect(splitAttributeAndFilterOperator('___some_name___ne')).toEqual({ attributeName: '___some_name_', operator: 'ne' })

    })


    it('should fail on wrong inputs', () => {


      function fn1() {
        splitAttributeAndFilterOperator()
      }

      function fn2() {
        splitAttributeAndFilterOperator([])
      }

      function fn3() {
        splitAttributeAndFilterOperator({ any: 'thing' })
      }

      function fn4() {
        splitAttributeAndFilterOperator('name__')
      }

      function fn5() {
        splitAttributeAndFilterOperator('name___')
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
      expect(fn4).toThrowErrorMatchingSnapshot();
      expect(fn5).toThrowErrorMatchingSnapshot();

    })

  })



  describe('transformFilterLevel', () => {


    it('should process filter level', () => {

      const goodFilter1 = {
        lastName: 'Doe',
        firstName__gte: 'J',
      }

      const result1 = {
        lastName: 'Doe',
        firstName: {
          $gte: 'J'
        }
      }


      const goodFilter2 = {
        lastName__in: ['Doe', 'Smith'],
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        isActive: true,
      }

      const result2 = {
        lastName: {
          $in: ['Doe', 'Smith'],
        },
        firstName: {
          $starts_with: 'Joh',
          $ends_with: 'an',
        },
        isActive: true,
      }


      expect(
        transformFilterLevel(goodFilter1, filteredEntity.getAttributes(), ['somewhere'])
      ).toEqual(result1)

      expect(
        transformFilterLevel(goodFilter2, filteredEntity.getAttributes(), ['somewhere'])
      ).toEqual(result2)

    })


    it('should process nested filter levels', () => {

      const goodFilter1 = {
        lastName: 'Doe',
        OR: [
          {
            firstName: 'Jack'
          },
          {
            firstName__starts_with: 'A',
            isActive: true
          }
        ]
      }

      const result1 = {
        lastName: 'Doe',
        $or: [
          {
            firstName: 'Jack'
          },
          {
            firstName: {
              $starts_with: 'A'
            },
            isActive: true
          }
        ]
      }


      const goodFilter2 = {
        lastName__gt: 'Tomson',
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        AND: [
          {
            lastName__starts_with: 'Und',
            lastName__ends_with: 'ton',
          }
        ],
        isActive: true,
      }

      const result2 = {
        lastName: {
          $gt: 'Tomson'
        },
        firstName: {
          $starts_with: 'Joh',
          $ends_with: 'an',
        },
        $and: [
          {
            lastName: {
              $starts_with: 'Und',
              $ends_with: 'ton',
            }
          }
        ],
        isActive: true,
      }


      expect(
        transformFilterLevel(goodFilter1, filteredEntity.getAttributes(), ['somewhere'])
      ).toEqual(result1)

      expect(
        transformFilterLevel(goodFilter2, filteredEntity.getAttributes(), ['somewhere'])
      ).toEqual(result2)

    })


    it('should throw if provided params are invalid', () => {

      function fn1() {
        transformFilterLevel('a')
      }

      function fn2() {
        transformFilterLevel([], null, [])
      }

      function fn3() {
        transformFilterLevel([], null, ['somewhere'])
      }

      function fn4() {
        transformFilterLevel([], null, ['somewhere', 'deeply', 'nested'])
      }

      function fn5() {
        transformFilterLevel({}, null, ['somewhere', 'deeply', 'nested'])
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
      expect(fn4).toThrowErrorMatchingSnapshot();
      expect(fn5).toThrowErrorMatchingSnapshot();

    })


    it('should throw if invalid attributes are used in filter', () => {

      const badFilter1 = {
        lastName: 'Doe',
        firstName__gte: 'J',
        something__ne: true,
      }

      const badFilter2 = {
        lastName: 'Doe',
        firstName__gte: 'J',
        anything_here: 'test',
      }


      function fn1() {
        transformFilterLevel(badFilter1, filteredEntity.getAttributes(), null)
      }

      function fn2() {
        transformFilterLevel(badFilter2, filteredEntity.getAttributes(), null)
      }

      function fn3() {
        transformFilterLevel(badFilter2, filteredEntity.getAttributes(), ['just', 'here'])
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    })



    it('should throw if exact match operators is used with another operator on the same attribute', () => {

      const badFilter1 = {
        lastName__in: ['Doe', 'Smith'],
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        firstName: 'Frank',
        isActive: true,
      }

      const badFilter2 = {
        lastName__in: ['Doe', 'Smith'],
        firstName: 'Frank',
        firstName__starts_with: 'Joh',
        firstName__ends_with: 'an',
        isActive: true,
      }

      function fn1() {
        transformFilterLevel(badFilter1, filteredEntity.getAttributes(), null)
      }

      function fn2() {
        transformFilterLevel(badFilter2, filteredEntity.getAttributes(), null)
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();

    })

  })

})
