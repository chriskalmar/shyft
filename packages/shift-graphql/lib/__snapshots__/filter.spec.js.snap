// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 1`] = `"invalid filter 'undefined'"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 2`] = `"invalid filter ''"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 3`] = `"invalid filter '[object Object]'"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 4`] = `"invalid filter 'name__'"`;

exports[`filter splitAttributeAndFilterOperator should fail on wrong inputs 5`] = `"invalid filter 'name___'"`;

exports[`filter transformFilterLevel should throw if exact match operators is used with another operator on the same attribute 1`] = `"Cannot combine 'exact match' operator with other operators on attribute 'firstName' used in filter"`;

exports[`filter transformFilterLevel should throw if exact match operators is used with another operator on the same attribute 2`] = `"Cannot combine 'exact match' operator with other operators on attribute 'firstName' used in filter"`;

exports[`filter transformFilterLevel should throw if invalid attributes are used in filter 1`] = `"Unknown attribute name 'something' used in filter"`;

exports[`filter transformFilterLevel should throw if invalid attributes are used in filter 2`] = `"Unknown attribute name 'anything_here' used in filter"`;

exports[`filter transformFilterLevel should throw if invalid attributes are used in filter 3`] = `"Unknown attribute name 'anything_here' used in filter at 'just.here'"`;

exports[`filter transformFilterLevel should throw if provided params are invalid 1`] = `"filter needs to be an object of filter criteria"`;

exports[`filter transformFilterLevel should throw if provided params are invalid 2`] = `"optional path in transformFilterLevel() needs to be an array"`;

exports[`filter transformFilterLevel should throw if provided params are invalid 3`] = `"filter at 'somewhere' needs to be an object of filter criteria"`;

exports[`filter transformFilterLevel should throw if provided params are invalid 4`] = `"filter at 'somewhere.deeply.nested' needs to be an object of filter criteria"`;

exports[`filter transformFilterLevel should throw if provided params are invalid 5`] = `"transformFilterLevel() expects an attribute map"`;
